src/components/TVChartContainer/TVChartContainer.tsx

```diff
@@ -0,0 +1,172 @@
+import { useCallback, useEffect, useRef, useState } from "react";
+import { TV_CHART_RELOAD_TIMESTAMP_KEY, TV_SAVE_LOAD_CHARTS_KEY } from "config/localStorage";
+import { useLocalStorage, useMedia } from "react-use";
+import { defaultChartProps, disabledFeaturesOnMobile } from "./constants";
+import useTVDatafeed from "domain/tradingview/useTVDatafeed";
+import { ChartData, IChartingLibraryWidget, IPositionLineAdapter } from "../../charting_library";
+import { getObjectKeyFromValue } from "domain/tradingview/utils";
+import { SaveLoadAdapter } from "./SaveLoadAdapter";
+import { SUPPORTED_RESOLUTIONS } from "config/tradingview";
+
+const TV_RELOAD_INTERVAL = 15 * 60 * 1000; // 15 minutes
+
+type ChartLine = {
+  price: number;
+  title: string;
+};
+
+type Props = {
+  symbol: string;
+  chainId: number;
+  savedShouldShowPositionLines: boolean;
+  chartLines: ChartLine[];
+  onSelectToken: () => void;
+  period: string;
+  setPeriod: (period: string) => void;
+};
+
+export default function TVChartContainer({
+  symbol,
+  chainId,
+  savedShouldShowPositionLines,
+  chartLines,
+  onSelectToken,
+  period,
+  setPeriod,
+}: Props) {
+  const chartContainerRef = useRef<HTMLDivElement | null>(null);
+  const tvWidgetRef = useRef<IChartingLibraryWidget | null>(null);
+  const [chartReady, setChartReady] = useState(false);
+  const [tvCharts, setTvCharts] = useLocalStorage<ChartData[] | undefined>(TV_SAVE_LOAD_CHARTS_KEY, []);
+  const { datafeed, resetCache } = useTVDatafeed();
+  const isMobile = useMedia("(max-width: 550px)");
+  const symbolRef = useRef(symbol);
+
+  const drawLineOnChart = useCallback(
+    (title: string, price: number) => {
+      if (chartReady && tvWidgetRef.current?.activeChart?.().dataReady()) {
```

Comment:
is it possible that chartReady == true but the second condition is false?

1. if yes, them useCallback deps should contain something related to tvWidgetRef.current.(etc)?
2. if no only using `chartReady` inside condition should be sufficient?

Reply:
Yes, it's possible. SO `chartReady` means if the chart is initially rendered. When wen change the token on the rendered chart the value of dataReady can change and be false.

Regarding putting the value of `useRef` i.e tvWidgetRef to the dependency array does't makes sense because `useRef` stores a mutable value. Please let me know if I am missing something.